kornia/contrib/dsnt.py

```diff
@@ -0,0 +1,220 @@
+
+    batch_size, channels, height, width = input.shape
+
+    # Create coordinates grid.
+    grid: torch.Tensor = create_meshgrid(
+        height, width, normalized_coordinates)
+    grid = grid.to(device=input.device, dtype=input.dtype)
+
+    pos_x: torch.Tensor = grid[..., 0].reshape(-1)
+    pos_y: torch.Tensor = grid[..., 1].reshape(-1)
+
+    input_flat: torch.Tensor = input.view(batch_size, channels, -1)
+
+    # Compute the expectation of the coordinates.
+    expected_y: torch.Tensor = torch.sum(pos_y * input_flat, -1, keepdim=True)
+    expected_x: torch.Tensor = torch.sum(pos_x * input_flat, -1, keepdim=True)
+
+    output: torch.Tensor = torch.cat([expected_x, expected_y], -1)
+
+    return output.view(batch_size, channels, 2)  # BxNx2
+
+
+def render_gaussian_2d(
+        mean: torch.Tensor,
+        std: torch.Tensor,
+        size: List[int],
+        normalized_coordinates: bool = True,
+):
+    r"""Renders the PDF of a 2D Gaussian distribution.
+
+    Arguments:
+        mean (torch.Tensor): the mean location of the Gaussian to render,
+          :math:`(\mu_x, \mu_y)`.
+        std (torch.Tensor): the standard deviation of the Gaussian to render,
+          :math:`(\sigma_x, \sigma_y)`.
+        size (list): the (height, width) of the output image.
+        normalized_coordinates: whether `mean` and `std` are assumed to use
+          coordinates normalized in the range of [-1, 1]. Otherwise,
+          coordinates are assumed to be in the range of the output shape.
+          Default is True.
+
+    Shape:
+        - `mean`: :math:`(*, 2)`
+        - `std`: :math:`(*, 2)`
+        - Output: :math:`(*, H, W)`
+    """
+    device = mean.device
+    dtype = mean.dtype
+    if not dtype.is_floating_point:
+        raise TypeError("Expected `mean` to have floating point values. Got {}"
+                        .format(dtype))
+    height, width = size
+
+    # Create coordinates grid.
+    grid: torch.Tensor = create_meshgrid(height, width, normalized_coordinates)
+    grid = grid.to(device=device, dtype=dtype)
+    pos_x: torch.Tensor = grid[..., 0].view(height, width)
+    pos_y: torch.Tensor = grid[..., 1].view(height, width)
+
+    # Gaussian PDF = exp(-(x - \mu)^2 / (2 \sigma^2))
+    #              = exp(dists * ks),
+    #                where dists = (x - \mu)^2 and ks = -1 / (2 \sigma^2)
+
+    # dists <- (x - \mu)^2
+    dist_x = (pos_x - mean[..., 0, None, None]) ** 2
+    dist_y = (pos_y - mean[..., 1, None, None]) ** 2
+
+    # ks <- -1 / (2 \sigma^2)
+    k_x = -0.5 * torch.reciprocal(std[..., 0, None, None])
+    k_y = -0.5 * torch.reciprocal(std[..., 1, None, None])
+
+    # Assemble the 2D Gaussian.
+    exps_x = torch.exp(dist_x * k_x)
+    exps_y = torch.exp(dist_y * k_y)
+    gauss = exps_x * exps_y
+
+    # Rescale so that values sum to one.
+    val_sum = gauss.sum(-2, keepdim=True).sum(-1, keepdim=True)
+    val_sum = val_sum + torch.finfo(dtype).tiny  # Avoid division by zero.
```

Comment:
it could be useful to abstract this to an external function, I think something like that might work
```python
def safe_zero_division(numerator: torch.Tensor, denominator: torch.Tensor) -> torch.Tensor:
    eps: float = torch.finfo(numerator.dtype).tiny
    return numerator / torch.clamp(denominator, min=eps)
```
/cc @bhack what do you think?

Reply:
I've broken it out into a local helper function. However, if you decide to make it a more accessible utility then it's important to make it crystal clear that it will only work properly for non-negative denominators.